### PR TITLE
Groth16 solidity verifier

### PR DESCRIPTION
**Pull request by JangoCCC**

### CODE REVIEW

File: groth16/src/api.rs [hunk 1]
```diff
@@ -148,6 +147,113 @@ pub fn groth16_verify(
 
     Ok(())
 }
+pub fn generate_verifier(vk_file_path: &str, sol_file_path: &str) -> Result<()> {
```
Comment (eigmax) [inline groth16/src/api.rs:150]: Can we declare that we adopt those code from Zokrate?

File: zkit/src/main.rs [hunk 2]
```diff
@@ -108,7 +109,9 @@ struct VerifyOpt {
 struct GenerateVerifierOpt {
     #[arg(short, default_value = "vk.bin")]
     vk_file: String,
-    #[arg(long = "s", default_value = "verifier.sol")]
+    #[arg(short, default_value = "groth16")]
```
Comment (eigmax) [inline zkit/src/main.rs:112]: Better to keep plonk as the default value

Comment (JangoCCC) [reply zkit/src/main.rs]: done

File: groth16/src/api.rs (hunk 1 above)
Comment (JangoCCC) [reply groth16/src/api.rs:152]: done

File: zkit/src/main.rs [hunk 3]
```diff
@@ -419,7 +422,20 @@ fn main() {
...
+        Command::GenerateVerifier(args) => {
+            match args.protocal.as_str() {
+                "plonk" => {
+                    plonky::api::generate_verifier(&args.vk_file, &args.sol)
+                }
+                "groth16" => {
+                    groth16::api::generate_verifier(&args.vk_file, &args.sol)
+                }
+                _ => {
+                    println!("unknown protocol");
```
Comment (eigmax) [inline zkit/src/main.rs:434]: you can use panic!("unknown protocol") to handle the exception.

Comment (JangoCCC) [reply zkit/src/main.rs]: done

File: test/single/hardhat.config.ts [hunk 4]
```diff
@@ -20,7 +20,7 @@ task("accounts", "Prints the list of accounts", async (taskArgs, hre) => {
  * @type import('hardhat/config').HardhatUserConfig
  */
 module.exports = {
-  solidity: "0.8.16",
+  solidity: "0.6.11",
```
Comment (eigmax) [inline test/single/hardhat.config.ts:23]: upgrade the solidity version to 0.8.16

File: test/single/input/groth16_proof.json [hunk 5]
```diff
@@ -0,0 +1,16 @@
+{
```
Comment (eigmax) [inline test/single/input/groth16_proof.json:1]: Remove this file.

File: groth16/src/constant.rs [hunk 6]
```diff
@@ -0,0 +1,54 @@
+pub const CONTRACT_TEMPLATE: &str = r#"
```
Comment (eigmax) [inline groth16/src/constant.rs:1]: this file is a template, not a constant.

File: test/single/input/groth16_public_input.json [hunk 7]
```diff
@@ -0,0 +1,3 @@
+[
```
Comment (eigmax) [inline test/single/input/groth16_public_input.json:1]: Remove this file.

File: test/test_groth16_verifier.sh [hunk 8]
```diff
@@ -0,0 +1,31 @@
+#!/bin/bash
+set -ex
```
Comment (eigmax) [inline test/test_groth16_verifier.sh:2]: move the command `generate_verifiler` to snark_verifier.sh, and remove this file.

**Pull request merged**